src/combinatorics/simple_graph/ends/defs.lean

```diff
@@ -0,0 +1,280 @@
+assigning, to each finite set of vertices, the connected components of its complement.
+-/
+
+
+universes u
+variables {V : Type u} (G : simple_graph V) (K L L' M : set V)
+
+open classical
+
+noncomputable theory
+local attribute [instance] prop_decidable
+
+namespace simple_graph
+
+section out
+
+/-- The graph induced by removing `K` -/
+@[reducible] def out := G.induce $ Kᶜ
+
+/-- Subsetship induces an obvious map on the induced graphs. -/
+@[reducible] def out_hom {K L} (h : K ⊆ L) : G.out L →g G.out K :=
+{ to_fun := λ ⟨v, hvK⟩, ⟨v, set.compl_subset_compl.mpr h hvK⟩,
+  map_rel' := by { rintros ⟨_, _⟩ ⟨_, _⟩, exact id } }
+
+lemma out_hom_refl (K) : G.out_hom (subset_refl K) = hom.id :=
+by { ext ⟨_, _⟩, refl, }
+
+lemma out_hom_trans {K L M} (h : K ⊆ L) (h' : L ⊆ M) :
+  G.out_hom (h.trans h') = (G.out_hom h).comp (G.out_hom h') :=
+by { ext ⟨_, _⟩, refl, }
+
+lemma out_hom_injective {K} {L} (h : K ⊆ L) : function.injective (G.out_hom h) :=
+by { rintros ⟨v, _⟩ ⟨w, _⟩ e,
+    simpa only [out_hom, subtype.mk_eq_mk, rel_hom.coe_fn_mk] using e, }
+
+end out
+
+/-- The components outside a given set of vertices `K` -/
+@[reducible] def comp_out := (G.out K).connected_component
+
+variables {G} {K L M}
+
+/-- The set of vertices of `G` making up the connected component `C` -/
+@[reducible, simp] def comp_out.supp (C : G.comp_out K) : set V :=
+  { v : V | ∃ (h : v ∈ K ᶜ), connected_component_mk (G.out K) ⟨v, h⟩ = C }
+
+@[ext] lemma comp_out.eq_iff_supp_eq (C D : G.comp_out K) : C = D ↔ C.supp = D.supp :=
+begin
+  split,
+  { rintro ⟨⟩, refl, },
+  { refine connected_component.ind₂ _ C D,
+    rintros ⟨v, hv⟩ ⟨w, hw⟩ h,
+    simp only [set.ext_iff, connected_component.eq, set.mem_compl_iff, set.mem_set_of_eq] at h ⊢,
+    exact ((h v).mp ⟨hv, reachable.refl _⟩).some_spec, }
+end
+
+instance : set_like (G.comp_out K) V :=
+{ coe := comp_out.supp,
+  coe_injective' := λ C D, (comp_out.eq_iff_supp_eq _ _).mpr, }
+
+namespace comp_out
+
+@[simp] lemma mem_supp_iff {v : V} {C : comp_out G K} :
+  v ∈ C ↔ ∃ (vK : v ∈ K ᶜ), connected_component_mk (G.out K) ⟨v, vK⟩ = C := iff.rfl
+
+/-- The connected component of `v`. -/
+@[reducible] def of_vertex (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) : G.comp_out K :=
+  connected_component_mk (G.out K) ⟨v, vK⟩
+
+lemma of_vertex_mem (G : simple_graph V) {v : V} (vK : v ∈ K ᶜ) :
+  v ∈ (comp_out.of_vertex G vK : G.comp_out K) := ⟨vK, rfl⟩
+
+lemma of_vertex_eq_of_adj (G : simple_graph V) {v w : V} (vK : v ∈ K ᶜ) (wK : w ∈ K ᶜ) :
+  G.adj v w → comp_out.of_vertex G vK = comp_out.of_vertex G wK :=
+by { rw [connected_component.eq], rintro a, apply adj.reachable, exact a }
+
+/-- The induced subgraph of `V` given by the vertices of `C`. -/
+@[reducible, protected]
+def subgraph (C : comp_out G K) : G.subgraph := (⊤ : G.subgraph).induce (C : set V)
+
+/-- The infinite connected components. -/
+@[reducible]
+def inf (C : G.comp_out K) := (C : set V).infinite
```

Comment:
This is a bad name. Do you really need this definition anyway?

Reply:
We can probably get rid of it